src/util/crypto_browser.ts

```diff
@@ -0,0 +1,135 @@
+// According to their type declarations, SubtleCrypto functions return `PromiseLike`.
+// They are close enough to a `Promise` to `await`, but tslint doesn't know that.
+/* tslint:disable:await-promise */
+
+const { crypto } = window
+const HASH_ALGORITHM = 'SHA-256'
+const ENCRYPTION_ALGORITHM = 'AES-GCM'
+const IV_LENGTH = 12
+const AUTH_TAG_BYTES = 16
+const AUTH_TAG_BITS = 8 * AUTH_TAG_BYTES
+const CACHE_EXPIRY = 30000
+
+// Cache keys so that `subtle.importKey` doesn't need to be called for every operation.
+// It would be nicer to just store the `CryptoKey`s on the stream `Connection`, but
+// that's tricky since this file takes the place of `crypto_node.ts`.
+class KeyCache {
+  private cache: Map<Buffer, CacheEntry> = new Map()
+
+  cleanup () {
+    const now = Date.now()
+    for (const entry of this.cache) {
+      const cacheData = entry[0]
+      const cacheEntry = entry[1]
+      if (now - cacheEntry.accessTime > CACHE_EXPIRY) {
+        this.cache.delete(cacheData)
+      }
+    }
+  }
+
+  async importKey (
+    keyData: Buffer,
+    algorithm: string | HmacImportParams | AesKeyAlgorithm,
+    keyUsages: string[]
+  ): Promise<CryptoKey> {
+    const oldEntry = this.cache.get(keyData)
+    if (oldEntry) {
+      oldEntry.accessTime = Date.now()
+      return oldEntry.keyObject
+    }
+    const keyObject = await crypto.subtle.importKey(
+      'raw',
+      keyData,
+      algorithm,
+      false, // extractable
+      keyUsages
+    )
+    this.cache.set(keyData, {
+      keyObject,
+      accessTime: Date.now()
+    })
+    return keyObject
+  }
+}
+
+interface CacheEntry {
+  keyObject: CryptoKey,
+  accessTime: number // milliseconds since epoch
+}
+
+const hmacKeyCache = new KeyCache()
+const aesKeyCache = new KeyCache()
+
+setInterval(() => {
```

Comment:
If the map gets big, this could potentially block the thread every 30s as it loops over the maps. Probably not an issue to deal with now.

Reply:
The map shouldn't ever be very big.